cub/test/catch2_test_device_histogram.cu

```diff
@@ -0,0 +1,662 @@
+}
+
+CUB_TEST("DeviceHistogram::Histogram* odd image sizes", "[histogram][device]")
+{
+  using sample_t                = int;
+  using level_t                 = int;
+  constexpr sample_t max_level  = 256;
+  constexpr int max_level_count = 256 + 1;
+
+  using P      = std::pair<int, int>;
+  const auto p = GENERATE(P{1920, 0}, P{0, 0}, P{0, 1080}, P{1, 1}, P{15, 1}, P{1, 15}, P{10000, 1}, P{1, 10000});
+  test_even_and_range<sample_t, 4, 3, int, level_t, int>(max_level, max_level_count, p.first, p.second);
+}
+
+CUB_TEST("DeviceHistogram::Histogram* entropy", "[histogram][device]")
+{
+  // TODO(bgruber): what numbers to put here? radix_sort_keys has GENERATE(1, 3, 9, 15);
+  const int entropy_reduction = GENERATE(-1, 5); // entropy_reduction = -1 -> all samples == 0
+  test_even_and_range<int, 4, 3, int>(256, 256 + 1, 1920, 1080, entropy_reduction);
+}
+
+template <int Channels, int ActiveChannels>
+struct ChannelConfig
+{
+  static constexpr auto channels        = Channels;
+  static constexpr auto active_channels = ActiveChannels;
+};
+
+CUB_TEST_LIST("DeviceHistogram::Histogram* channel configs",
+              "[histogram][device]",
+              ChannelConfig<1, 1>,
+              ChannelConfig<3, 3>,
+              ChannelConfig<4, 3>,
+              ChannelConfig<4, 4>)
+{
+  test_even_and_range<int, TestType::channels, TestType::active_channels, int, int, int>(256, 256 + 1, 128, 32);
+}
+
+CUB_TEST("DeviceHistogram::HistogramEven sample iterator", "[histogram_even][device]")
+{
+  using sample_t                = int;
+  const auto width              = 100;
+  const auto padding            = 13; // in elements
+  const auto height             = 30;
+  constexpr auto Channels       = 4;
+  constexpr auto ActiveChannels = 3;
+  const auto row_pitch          = (width + padding) * Channels * static_cast<int>(sizeof(sample_t));
+  const auto total_values       = (width + padding) * Channels * height;
+
+  const auto num_levels  = array<int, ActiveChannels>{11, 3, 2};
+  const auto lower_level = array<int, ActiveChannels>{0, -10, std::numeric_limits<int>::lowest()};
+  const auto upper_level = array<int, ActiveChannels>{total_values, 10, std::numeric_limits<int>::max()};
+
+  auto sample_iterator = cub::CountingInputIterator<sample_t>(0);
+
+  // Channel #0: 0, 4,  8, 12
+  // Channel #1: 1, 5,  9, 13
+  // Channel #2: 2, 6, 10, 14
+  // unused:     3, 7, 11, 15
+
+  auto d_histogram = array<c2h::device_vector<int>, ActiveChannels>();
+  for (int c = 0; c < ActiveChannels; ++c)
+  {
+    d_histogram[c].resize(num_levels[c] - 1);
+  }
+
+  multi_histogram_even<Channels, ActiveChannels>(
+    sample_iterator,
+    to_ptr_array(d_histogram).data(),
+    num_levels.data(),
+    lower_level.data(),
+    upper_level.data(),
+    width,
+    height,
+    row_pitch);
+
+  // TODO(bgruber): move detail::to_vec into namespace c2h?
+  CHECK(detail::to_vec(d_histogram[0]) == std::vector<int>(10, (width * height) / 10));
```

Comment:
I think `to_vec` would be useful outside the `detail` namespace. It converts host/device vectors to `std::vector`.

Reply:
Obsolete.